### ISSUE TITLE
Realms in which display of Bidi data is not guided by UBA ( i.e. zOS, Adobe Reader)

### ISSUE THREAD

**Issue by tomerm** [w3c/alreq#119]
Several concepts associated with bidirectional text display (i.e. those presented in https://github.com/w3c/alreq/wiki/Draft-for-%E2%80%9C2.3-Direction%E2%80%9D) are highly dependent on underlying operating system (i.e. Windows vs zOS) or application (i.e. Notepad vs Adobe Reader).
Specifically UBA is not always used for presenting text on the screen. This is true for operating systems (such as zOS - https://en.wikipedia.org/wiki/Z/OS) and also applications (such as Adobe Reader).

Software running or/and interacting with data coming from mainframe (a computer running zOS) is coming from many companies (not only from IBM). For example:

- http://www.rocketsoftware.com/products?field_product2_platforms_tid=381&rcid=mm-pr-pf-zsystems
- http://www.pacsys.com/
- https://www.emtec.com/zoc/tn3270-terminal-emulation.html
- https://www.mochasoft.dk/tn3270.htm
- http://www.sdisw.com/
- http://bosanova.net/

Do you believe it makes sense to at least briefly mention the existence of systems / applications in which bidirectional data display is not guided by UBA ? 

PS. This is by the way one of the reasons why indexing Arabic PDFs is a challenging task (as @khaledhosny mentioned in #107).

**Comment by khaledhosny** [comment]
I have no idea about zOS, but PDF does not contain text but positioned glyphs, so UBA is out of scope as it has to be done by the PDF creators not the readers.

The issue with indexing PDFs is unrelated, but it also needs to be fixed mainly in PDF creators.

**Comment by shervinafshar** [comment]
I agree with Khaled. Our focus is mainly on text layout rather than mere display of Arabic text (as it may happen with any software system storing or interchanging Arabic). Even advanced text editors which are not intended for layouting text might decided to ignore UBA or any other UTR or UAX for the sake of simplicity; e.g UAX#14 is an interesting case to test in your favorite text editor.

**Comment by tomerm** [comment]
@khaledhosny @shervinafshar thanks a lot for your comments. 
I think one of the key point to understand / clarify  is if as part of alreq you provide a general guidance for Arabic text layout (for example on a piece of paper, book, sand, water, piece of metal, stone or wood etc.) or the guidance is tailored to computer based world. Based on _".... This document describes requirements for the layout and presentation of text in languages that use the Arabic script when they are used by Web standards and technologies, such as HTML, CSS, Mobile Web, Digital Publications, and Unicode...."_  https://w3c.github.io/alreq/ I would presume that you target web technologies. This perception is also indirectly supported by the fact that UBA is mentioned in section 2.3 https://w3c.github.io/alreq/#h_direction

I also believe (please correct me if I am wrong) that the guidance provided in https://w3c.github.io/alreq/ is meant to target (in particular) the audience of software developers / product owners to assure their products address requirements associated with Arabic text layout. 

No doubt UBA is very widely spread standard leveraged by many ( I would even say overwhelming majority ) software product / platforms. As compared to it zOS terminals, Adobe Reader and text editor which decides to ignore UBA are in minority. I thus don't believe that a lot of attention should be dedicated in the guidance to edge cases like that. Having said that I do believe that it should be at least briefly mentioned that while UBA is widely spread standard, there are several cases in which it is not used for laying out text on the screen. For a reference please see this article: https://www.w3.org/International/questions/qa-visual-vs-logical 
If you still believe that none of such comments are appropriate please feel free to close this issue as irrelevant. 

Coming back to PDF / zOS for the sake of clarification only. As @khaledhosny mentioned, in PDF case 
the responsibility of text reordering is in the hands of text authors (PDF creators) rather than in the hands of text readers / viewers. This is exactly what happens with zOS terminals emulators. It is text author who should run UBA in his / her head (so to speak) and write it in already reordered form. It is very similar to how you write text on the paper. You reorder it in your head, move your hand to the expected spot above the paper and write down the letter using a pen. 

As opposed to this way, modern systems don't expect you run any reordering in your head. You simply type the text in the same order / way you pronounce it. The rest is done by UBA (employed by platform / application etc.). Namely UBA reorders this text before it is rendered on the display. 

Now, imagine you exchange text as-is between those two worlds. For example, you extract text from PDF (as-is), put it into HTML file (as-is) and upload the file to a web browser. Assuming text in PDF was correctly displayed in Acrobat Reader, it won't be properly displayed in web browser. This is since web browser assumes text is provided not reordered and will apply UBA on it in order to reorder it. However, text as-is extracted from PDF is already reordered. As a result text won't be displayed correctly in web since it was reordered twice (first time as part of text authoring during PDF content creation, second time when after being extracted from PDF it is rendered inside web browser).
